packages/library/src/compute-limit-internal.ts

```diff
@@ -0,0 +1,187 @@
+ * > or a multiplier.
+ *
+ * > [!NOTE]
+ * > If you are preparing an _unsigned_ transaction, destined to be signed and submitted to the
+ * > network by a wallet, you might like to leave it up to the wallet to determine the compute unit
+ * > limit. Consider that the wallet might have a more global view of how many compute units certain
+ * > types of transactions consume, and might be able to make better estimates of an appropriate
+ * > compute unit budget.
+ */
+export async function getComputeUnitEstimateForTransactionMessage_INTERNAL_ONLY_DO_NOT_EXPORT({
+    abortSignal,
+    rpc,
+    transactionMessage,
+    ...simulateConfig
+}: ComputeUnitEstimateForTransactionMessageConfig): Promise<number> {
+    /**
+     * STEP 1: Make sure the transaction message will not fail in simulation for lack of a lifetime
+     *         - either a recent blockhash lifetime or a nonce.
+     */
+    const isDurableNonceTransactionMessage = isDurableNonceTransaction(transactionMessage);
+    let compilableTransactionMessage;
+    if (isDurableNonceTransactionMessage || isTransactionMessageWithBlockhashLifetime(transactionMessage)) {
+        compilableTransactionMessage = transactionMessage;
+    } else {
+        compilableTransactionMessage = setTransactionMessageLifetimeUsingBlockhash(
+            INVALID_BUT_SUFFICIENT_FOR_COMPILATION_BLOCKHASH,
+            transactionMessage,
+        );
+    }
+    /**
+     * STEP 2: Ensure that the message has a `SetComputeLimit` instruction. The set compute limit
+     *         instruction itself consumes compute units, so it must be included in the simulation.
+     */
+    const existingSetComputeUnitLimitInstructionIndex =
+        transactionMessage.instructions.findIndex(isSetComputeLimitInstruction);
+    const maxComputeUnitLimitInstruction = createComputeUnitLimitInstruction(4294967295 /* U32::MAX */);
+    if (existingSetComputeUnitLimitInstructionIndex === -1) {
+        compilableTransactionMessage = appendTransactionMessageInstruction(
+            maxComputeUnitLimitInstruction,
+            compilableTransactionMessage,
+        );
+    } else {
+        const nextInstructions = [...compilableTransactionMessage.instructions];
+        nextInstructions.splice(existingSetComputeUnitLimitInstructionIndex, 1, maxComputeUnitLimitInstruction);
+        compilableTransactionMessage = Object.freeze({
+            ...compilableTransactionMessage,
+            instructions: nextInstructions,
+        } as typeof compilableTransactionMessage);
+    }
+    /**
+     * STEP 3: Simulate the transaction to measure its compute unit consumption.
+     */
+    const compiledTransaction = compileTransaction(compilableTransactionMessage);
+    const wireTransactionBytes = getBase64EncodedWireTransaction(compiledTransaction);
+    try {
+        const {
+            value: { unitsConsumed },
+        } = await rpc
+            .simulateTransaction(wireTransactionBytes, {
+                ...simulateConfig,
```

Comment:
Since this is `library` and we can be opinionated, should we default the commitment to `Confirmed` instead of simulate's default of `Finalized`? It's most likely how they'll eventually send the transaction, and can still be explicitly set if needed.